adafruit_ble/advertising/__init__.py

```diff
@@ -0,0 +1,287 @@
+        data[i + 1 + key_size: i + 1 + item_length] = bytes(value)
+        i += 1 + item_length
+    return data
+
+class AdvertisingDataField:
+    """Top level class for any descriptor classes that live in Advertisement or it's subclasses."""
+
+class AdvertisingFlag:
+    """A single bit flag within an AdvertisingFlags object."""
+    def __init__(self, bit_position):
+        self._bitmask = 1 << bit_position
+
+    def __get__(self, obj, cls):
+        return (obj.flags & self._bitmask) != 0
+
+    def __set__(self, obj, value):
+        if value:
+            obj.flags |= self._bitmask
+        else:
+            obj.flags &= ~self._bitmask
+
+class AdvertisingFlags(AdvertisingDataField):
+    """Standard advertising flags"""
+
+    limited_discovery = AdvertisingFlag(0)
+    """Discoverable only for a limited time period."""
+    general_discovery = AdvertisingFlag(1)
+    """Will advertise until discovered."""
+    le_only = AdvertisingFlag(2)
+    """BR/EDR not supported."""
+    # BR/EDR flags not included here, since we don't support BR/EDR.
+
+    def __init__(self, advertisement, advertising_data_type):
+        self._advertisement = advertisement
+        self._adt = advertising_data_type
+        self.flags = None
+        if self._adt in self._advertisement.data_dict:
+            self.flags = self._advertisement.data_dict[self._adt][0]
+        elif self._advertisement.mutable:
+            self.flags = 0b110 # Default to General discovery and LE Only
+        else:
+            self.flags = 0
+
+    def __bytes__(self):
+        encoded = bytearray(1)
+        encoded[0] = self.flags
+        return encoded
+
+    def __str__(self):
+        parts = ["<AdvertisingFlags"]
+        for attr in dir(self.__class__):
+            attribute_instance = getattr(self.__class__, attr)
+            if issubclass(attribute_instance.__class__, AdvertisingFlag):
+                if getattr(self, attr):
+                    parts.append(attr)
+        parts.append(">")
+        return " ".join(parts)
+
+class String(AdvertisingDataField):
+    """UTF-8 encoded string in an Advertisement.
+
+       Not null terminated once encoded because length is always transmitted."""
+    def __init__(self, *, advertising_data_type):
+        self._adt = advertising_data_type
+
+    def __get__(self, obj, cls):
+        if self._adt not in obj.data_dict:
+            return None
+        return str(obj.data_dict[self._adt], "utf-8")
+
+    def __set__(self, obj, value):
+        obj.data_dict[self._adt] = value.encode("utf-8")
+
+class Struct(AdvertisingDataField):
+    """`struct` encoded data in an Advertisement."""
+    def __init__(self, struct_format, *, advertising_data_type):
+        self._format = struct_format
+        self._adt = advertising_data_type
+
+    def __get__(self, obj, cls):
+        if self._adt not in obj.data_dict:
+            return None
+        return struct.unpack(self._format, obj.data_dict[self._adt])[0]
```

Comment:
Something you may want to consider doing here is caching these values if they're expensive to compute. It's a memory/cpu tradeoff, though.

Reply:
Ya, perhaps. I think this is fine for now.